lib/_stream_writable.js

```diff
@@ -590,6 +591,8 @@ Writable.prototype.end = function(chunk, encoding, cb) {
   // Ignore unnecessary end() calls.
   if (!state.ending)
     endWritable(this, state, cb);
+  else if (typeof cb === 'function')
+    process.nextTick(cb, new ERR_STREAM_ENDED('end'));
```

Comment:
Can we document this behavior? It is currently documented as follows:

> If provided, the optional callback function is attached as a listener for the 'finish' event.

So I personally expect the callback to not be called with an error if the `'finish'` event has not been emitted yet.

Reply:
I concur with @lpinca (see my previous comment). If `'finish'` has not been emitted yet I prefer the cb to be added to the `'finish'` event.